rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,770 @@
+        assert 0 <= tag < 256
+
+        # Construct arg1 (op_code << 8) | app_id
+        arg1 = consts.AllocOperations.alloc_sdram << 8 | app_id
+
+        # Send the packet and retrieve the address
+        rv = self._send_scp(x, y, 0, SCPCommands.alloc_free, arg1, size, tag)
+        if rv.arg1 == 0:
+            # Allocation failed
+            raise SpiNNakerMemoryError(size, x, y)
+        return rv.arg1
+
+    @ContextMixin.use_contextual_arguments
+    def sdram_alloc_as_io(self, size, tag=0, x=Required, y=Required,
+                          app_id=Required):
+        """Like :py:meth:`.sdram_alloc` but a file-like object which allows
+        reading and writing to the region is returned.
+
+        Returns
+        -------
+        :py:class:`~MemoryIO`
+            File-like object which allows accessing the newly allocated region
+            of memory.
+
+        Raises
+        ------
+        SpiNNakerMemoryError
+            If the memory cannot be allocated, or the tag is already taken or
+            invalid.
+        """
+        # Perform the malloc
+        start_address = self.sdram_alloc(size, tag, x, y, app_id)
+
+        return MemoryIO(self, x, y, start_address, start_address + size)
+
+    def _get_next_nn_id(self):
+        """Get the next nearest neighbour ID."""
+        self._nn_id = self._nn_id + 1 if self._nn_id < 126 else 1
+        return self._nn_id * 2
+
+    def _send_ffs(self, pid, region, n_blocks, fr):
+        """Send a flood-fill start packet."""
+        sfr = fr | (1 << 31)
+        self._send_scp(
+            0, 0, 0, SCPCommands.nearest_neighbour_packet,
+            (NNCommands.flood_fill_start << 24) | (pid << 16) |
+            (n_blocks << 8),
+            region, sfr
+        )
+
+    def _send_ffd(self, pid, aplx_data, address):
+        """Send flood-fill data packets."""
+        block = 0
+        while len(aplx_data) > 0:
+            # Get the next block of data, send and progress the block
+            # counter and the address
+            data, aplx_data = (aplx_data[:consts.SCP_DATA_LENGTH],
+                               aplx_data[consts.SCP_DATA_LENGTH:])
+            size = len(data) // 4 - 1
+
+            arg1 = (NNConstants.forward << 24 | NNConstants.retry << 16 | pid)
+            arg2 = (block << 16) | (size << 8)
+            self._send_scp(0, 0, 0, SCPCommands.flood_fill_data,
+                           arg1, arg2, address, data)
+
+            # Increment the address and the block counter
+            block += 1
+            address += len(data)
+
+    def _send_ffe(self, pid, app_id, app_flags, cores, fr):
+        """Send a flood-fill end packet."""
+        arg1 = (NNCommands.flood_fill_end << 24) | pid
+        arg2 = (app_id << 24) | (app_flags << 18) | (cores & 0x3fff)
+        self._send_scp(0, 0, 0, SCPCommands.nearest_neighbour_packet,
+                       arg1, arg2, fr)
+
+    @ContextMixin.require_named_contextual_arguments("app_id", "app_flags")
```

Comment:
I would be tempted to make app_flags have a default of 0. Unless you're being clever, you don't need any flags. (For example, most users probably won't use the wait flag and then send SIG_START because their application already has a soft barrier on SYNC0).